### ISSUE TITLE
Constificator does not always detect read and write access

### ISSUE THREAD

**Issue by Darnor** [Cevelop/Issues#15]
### Expected Behavior

Given:

```
// Object.hpp
#ifndef OBJECT_HPP_
#define OBJECT_HPP_

struct Object {
	int n { }; // no warning
};

#endif
```

```
// main.cpp
#include "Object.hpp"

int main() {
	Object obj { };
	obj.n = 3; // write on n
}
```
No warning for `int n { }` should occur due to the write inside `main`. 

But given:
```
// main.cpp
struct Object {
	int n { }; // warning, this should be const
};

int main() {
	Object obj { }; // warning, this should be const
	int x { obj.n };
	x++;
}
```
A warning should occure for both `Object obj { }` and `int n { }` inside `struct Object`.

### Actual Behavior

In the first example `int n` inside `struct Object` get marked by the Constificator. This does not happen if the code resides inside the same file though:
```
// main.cpp
struct Object {
	int n { }; // no warning, which is correct.
};

int main() {
	Object obj { };
	obj.n = 42;
}
```
The warning inside `Object` happens only if `int n` gets initialized:
```
struct Object {
	int n; // no warning
};
```

In the second example no warning is given for neither `Object obj { };` and `int n { }` even though only read access occurs.

### Cevelop Version, Operating System and Compiler

Cevelop 1.6 with Constificator 1.6.0.201703201353

**Comment by Darnor** [comment]
No sure if this is a different issue but I just want to add that there is something similar going on while using filestreams.

Given:
```
auto readFile(
		std::string const & filename
	) -> std::string {

	std::ifstream fileStream { filename };
	std::string fileString { std::istreambuf_iterator<char> { fileStream }, std::istreambuf_iterator<char> { } };
	return fileString;
}
```
There are two things wrong here. Firstly the constificator warns that `std::ifstream fileStream { filename }` should be const, which it cannot since it gets modified while creating the output string. Secondly no warning is given that the fileString should actually be const due to no write access occuring.

Updating the code as follows will result in the correct behaviour, meaning no warning from the constificator for the `std::ifstream fileStream { filename }`.
```
auto readFile(
		std::string const & filename
	) -> std::string {

	std::ifstream fileStream { filename };
	return { std::istreambuf_iterator<char> { fileStream }, std::istreambuf_iterator<char> { } };
```

**Comment by PeterSommerlad** [comment]
In the example regarding variable `filestring` must not be `const`, because this would prohibit return-value-optimization.

**Comment by tcorbat** [comment]
> In the example regarding variable filestring must not be const, because this would prohibit return-value-optimization.

I object: [class.copy.elision] (1, 1) ... when the expression is the name of
a **non-volatile** automatic object (...) with the same type **(ignoring cv-qualification)** as the function
return type...

**Comment by PeterSommerlad** [comment]
see http://stackoverflow.com/questions/14429988/can-a-c-compiler-perform-rvo-for-a-const-return-value
I said RVO, but meant move-semantics for the return value.

**Comment by fmorgner** [comment]
@Darnor The original issue is fixed on the master branch. I looking into the second one. Thanks for reporting!

**Comment by fmorgner** [comment]
The second issue is fixed and scheduled for release 1.7.0. Thanks again for reporting!
